### PR TITLE
[Gutenberg] Unsupported block fallback

### PR DESCRIPTION
**Pull request by marecar3**
Proof of concept Unsupported block fallback on native side.
This opens a web view with a new-post editor, and insert the required block via "User Content Javascript" (JS code inserted to the web view).

We extract the final block content in the same way.

To clean up the canvas from unwanted UI elements we inject CSS that hides those elements. 

`gutenberg-mobile` side PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2063

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by guarani** [comment]
@marecar3 the downloadable build task failed so I haven't been able to test this.

**Comment by mchowning** [comment]
Most likely pulling in the latest changes (with the no-jitpack updates) will get this passing CI.

**Comment by marecar3** [comment]
> Most likely pulling in the latest changes (with the no-jitpack updates) will get this passing CI.

thanks @mchowning!
@guarani I have updated with the latest develop.

**Comment by oguzkocer** [comment]
Hi there! `14.8` is being frozen, so this PR is being bumped to `14.9`. If these changes _have to_ be shipped with `14.8`, please target `release/14.8` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

**Comment by guarani** [comment]
> I have updated with the latest develop.

Thanks @marecar3!

**Comment by guarani** [comment]
Hey @marecar3, I installed the apk and got this error upon opening the editor:

> Unable to load script. Make sure you are either running a Metro server (run 'react-native start') or that your bundle 'index.android.bundle' is packaged correctly for release.

I thought that if there was a problem with the React Native bundle, there was a UI test that would have caught that, so I'm not sure if it's an issue on my end or not.

**Comment by marecar3** [comment]
> Hey @marecar3, I installed the apk and got this error upon opening the editor:
> 
> > Unable to load script. Make sure you are either running a Metro server (run 'react-native start') or that your bundle 'index.android.bundle' is packaged correctly for release.
> 
> I thought that if there was a problem with the React Native bundle, there was a UI test that would have caught that, so I'm not sure if it's an issue on my end or not.

Yes you are right I am getting the same message. Maybe @hypest can help us?

**Comment by marecar3** [comment]
> > Hey @marecar3, I installed the apk and got this error upon opening the editor:
> > > Unable to load script. Make sure you are either running a Metro server (run 'react-native start') or that your bundle 'index.android.bundle' is packaged correctly for release.
> > 
> > 
> > I thought that if there was a problem with the React Native bundle, there was a UI test that would have caught that, so I'm not sure if it's an issue on my end or not.
> 
> Yes you are right I am getting the same message. Maybe @hypest can help us?

@guarani I have update the PR and now you could continue with testing. Thanks.

**Comment by guarani** [comment]
I played with a couple of blocks and I think it's working great, @marecar3!

I think the editing experience is good, although there's likely more scenarios I haven't covered yet. After testing on Android now and iOS earlier, the experience on both platforms is very much the same from what I can see – which is great news.

I tested the following blocks (chosen at random): Twitter embed, CoBlocks Offset Gallery, Grow Form

#### Scenarios tested
1. ✅ Edited a embedded tweet block URL and caption and verified result on the web
2. ✅ Transformed a Twitter block into a Paragraph block, went back to the editor and verified the the Paragraph block was shown correctly
3. ✅ Transformed one unsupported block into another: Transformed a CoBlocks Offset Gallery to CoBlocks Stack Gallery and verified results in the app editor and then on the web
4. ✅ Created a CoBlocks Offset Gallery block on the web and tweaked some settings in the app

#### Things that didn't look right

##### Element flashing on screen
As the web editor is loading, an element flashes on the screen for a split-second. My guess is that the code that hides unwanted UI elements is getting loaded too late ([9s](https://cldup.com/ZsbhbEumZT.mp4)), can this be changed @marecar3?

##### Colored status bar background
There's a blue/violet background behind the status bar (see the video in the above item) which seems out of place.

##### Save button confusion 
The Save button in the web view makes me think the changes are being saved ([had this same issue on iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/13967#issuecomment-622068654)) – @iamthomasbishop do you recall if we were going to change the web editor's save button title? It's currently "Save" which is confusing (at least for me).

##### User friendly names for all blocks

| Placeholder block | Bottom sheet |
| -- | -- |
|  |  |

Blocks such as the Offset Gallery don't have user friendly names on the Unsupported Blocks bottom sheet (e.g. "coblocks/gallery-offset isn't yet supported on WordPress for Android") – [we had this same issue on iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/13967#issuecomment-622068654). Is this a blocker for releasing unsupported blocks, @iamthomasbishop?

##### Discoverability of unsupported block editing



On the blank block we use in the editor as a placeholder for unsupported blocks, perhaps we should make it easier to find the editing options. Currently on Android you have to tap the question mark (?) – probably the same on iOS – to reveal the bottom sheet. Tapping anywhere else on the placeholder block doesn't currently reveal any options. This might also be for @iamthomasbishop - do you think this needs attention?

---

cc @etoledom

**Comment by iamthomasbishop** [comment]
> There's a blue/violet background behind the status bar (see the video in the above item) which seems out of place.

I agree w/ @guarani here, this is something that stood out to me when watching one of the testing videos he shared above. Can we force a white status bar (with black or dark labels/icons).

> do you recall if we were going to change the web editor's save button title? It's currently "Save" which is confusing (at least for me).

I think the discussed label on iOS was `Apply`, but not sure if that's final. On Android, we could probably change the `Save` button to a checkmark icon and the `<-` arrow to a `X` icon to indicate discard and apply. Would this align well with iOS, @etoledom?

> perhaps we should make it easier to find the editing options

This was an intentional design choice made before we knew what the external editing function would look like, but I think the same criteria still apply. Typically, we only make the entire block perform an action when there is an explicit action button (eg. the blue text button on blocks like image, video, latest posts, etc). The `(?)` button is a secondary action so it shouldn't (as a guideline) perform this primary action. I don't know if we should necessarily make this a _primary_ action, so my proposal would be to leave it.

I can't believe I haven't thought of this sooner but in general, for unsupported blocks, we could (and maybe should) show the settings/gear icon on the Inline Toolbar. When this is tapped could show this same sheet. // thoughts @etoledom ?

Another random idea: maybe if there is no primary action button, we should "bounce" or otherwise highlight the secondary action when the static placeholder is tapped in order to draw some attention to it (in this case, the `(?)` icon), although I'm not sure if that would feel odd in practice, it might at least aid in discovery.

**Comment by etoledom** [comment]
> I think the discussed label on iOS was Apply, but not sure if that's final.

We decided to go with `Continue` and `Cancel` for iOS.

> we could probably change the Save button to a checkmark icon and the <- arrow to a X. Would this align well with iOS, @etoledom?

We had an `x` button implemented on iOS, [but decided to go with `Cancel`](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2063#issuecomment-622106212) to be explicit about what this action is about.

The check mark sounds good to me ✔️ 

> Element flashing on screen

@marecar3 - On iOS there's an option on WebViews to disable incremental rendering of the page. In this way nothing is rendered until the page is fully loaded. Maybe there's a similar option on Android 🤔 

> User friendly names for all blocks

I understand that these internal names of blocks are not the best to show to the user, but it's what we got. The only option I see is to manually remove anything prefixed by `/`, replace `-` with spaces and (maybe) capitalise every word. @marecar3 @Guarani what do you think about this option? And @iamthomasbishop - How bad do you think these internal block names are? Could this wait for a second release?

> for unsupported blocks, we could (and maybe should) show the settings/gear icon on the Inline Toolbar.

Sounds good! How important do you think this is to include it on this first release iteration? I'd pretty much like to not add new features at this stage, but we can if really needed.

**Comment by marecar3** [comment]
Hey @cameronvoell 👋 
Could you be able to do a few iterations of code review on this PR? thanks.

**Comment by marecar3** [comment]
> I played with a couple of blocks and I think it's working great, @marecar3!
> 
> I think the editing experience is good, although there's likely more scenarios I haven't covered yet. After testing on Android now and iOS earlier, the experience on both platforms is very much the same from what I can see – which is great news.
> 
> I tested the following blocks (chosen at random): Twitter embed, CoBlocks Offset Gallery, Grow Form
> 
> #### Scenarios tested
> 1. ✅ Edited a embedded tweet block URL and caption and verified result on the web
> 2. ✅ Transformed a Twitter block into a Paragraph block, went back to the editor and verified the the Paragraph block was shown correctly
> 3. ✅ Transformed one unsupported block into another: Transformed a CoBlocks Offset Gallery to CoBlocks Stack Gallery and verified results in the app editor and then on the web
> 4. ✅ Created a CoBlocks Offset Gallery block on the web and tweaked some settings in the app
> 
> #### Things that didn't look right
> ##### Element flashing on screen
> As the web editor is loading, an element flashes on the screen for a split-second. My guess is that the code that hides unwanted UI elements is getting loaded too late ([9s](https://cldup.com/ZsbhbEumZT.mp4)), can this be changed @marecar3?
> 
> ##### Colored status bar background
> There's a blue/violet background behind the status bar (see the video in the above item) which seems out of place.
> 
> ##### Save button confusion
> The Save button in the web view makes me think the changes are being saved ([had this same issue on iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/13967#issuecomment-622068654)) – @iamthomasbishop do you recall if we were going to change the web editor's save button title? It's currently "Save" which is confusing (at least for me).
> 
> ##### User friendly names for all blocks
> Placeholder block	Bottom sheet
> 	
> Blocks such as the Offset Gallery don't have user friendly names on the Unsupported Blocks bottom sheet (e.g. "coblocks/gallery-offset isn't yet supported on WordPress for Android") – [we had this same issue on iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/13967#issuecomment-622068654). Is this a blocker for releasing unsupported blocks, @iamthomasbishop?
> 
> ##### Discoverability of unsupported block editing
> 
> 
> On the blank block we use in the editor as a placeholder for unsupported blocks, perhaps we should make it easier to find the editing options. Currently on Android you have to tap the question mark (?) – probably the same on iOS – to reveal the bottom sheet. Tapping anywhere else on the placeholder block doesn't currently reveal any options. This might also be for @iamthomasbishop - do you think this needs attention?
> 
> cc @etoledom

Hey, @guarani @iamthomasbishop I think that I have managed to address all issues that were mention in the comment. Could you do another iteration of testing? thanks :)

**Comment by guarani** [comment]
@marecar3 I took the liberty of updating this to 15.1 since 15.0 is being cut today.

**Comment by etoledom** [comment]
> For some content in the jetpack/markdown, the Unsupported fallback on iOS and Android shows "This block contains unexpected or invalid content" 

Great catch @cameronvoell !

What is happening is that the `\n` from `{"source":"### This is Markdown\n*Test*"}` is being transformed to an actual new line, breaking this property.

I'm not sure why is this happening, on iOS we are escaping it in the html string we pass to JS (like `\\n`) but on the JS side is received with `*Test*}` written on a new line.

Clearly it's happening on both platforms, so the solution might be on the Block insertion script (somehow). Will give it a try!

**Comment by etoledom** [comment]
> For some content in the jetpack/markdown, the Unsupported fallback on iOS and Android shows "This block contains unexpected or invalid content"

So, I haven’t been able to find a good solution for this, but I can confirm that the problem is what I mentioned previously.

The only thing that worked was to replace instances of `\n` with `\\\n` on the block html on native side before sending it to JS.

Of course this looks quite ugly and it might break blocks with new lines on its content.

I’m open to suggestions.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/WPGutenbergWebViewActivity.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public static final String ARG_AUTHENTICATION_TOKEN = "authenticated_token";
+    public static final String ARG_URL_TO_LOAD = "url_to_load";
+    public static final String ARG_IS_SITE_PRIVATE = "is_site_private";
+
+    public static final String ARG_BLOCK_ID = "block_id";
+    public static final String ARG_BLOCK_NAME = "block_name";
+    public static final String ARG_BLOCK_CONTENT = "block_content";
+
+    public static final String RESULT_SAVED_CONTENT = "saved_content";
+    public static final String RESULT_BLOCK_ID = "block_id";
```
[the remark this answers is not in the document]
Comment (marecar3) [reply libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/WPGutenbergWebViewActivity.java:29]: Agree. Fixed (we are now using only one type of param)

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/WPGutenbergWebViewActivity.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            return String.format(getString(R.string.menu_toolbar_title), blockName);
+        }
+        return "";
+    }
+
+    @Override
+    protected void saveContent(String content) {
+        Intent savedContentIntent = new Intent();
+        String blockId = getIntent().getExtras().getString(ARG_BLOCK_ID);
+        savedContentIntent.putExtra(ARG_BLOCK_ID, blockId);
```
[the remark this answers is not in the document]
Comment (marecar3) [reply libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/WPGutenbergWebViewActivity.java:60]: Agree. Fixed (we are now using only one type of param)

**Comment by marecar3** [comment]
> Thanks for the changes @marecar3, the changes in the PR look good. We may need some coordination with this PR which seems to be blocking us building with latest gutenberg-mobile develop branch #12041. But otherwise these changes are good to go.

Thanks @cameronvoell 🙇

**Comment by guarani** [comment]
> The only thing that worked was to replace instances of \n with \\\n on the block html on native side before sending it to JS.

That's true—what wasn't immediately obvious to me is that the content in the JSON object's `source` had to match the content in the `<p>` tag, otherwise I'd still get the "This block contains unexpected or invalid content" error. So I had to add one backslash inside the JSON and two more inside the `<p>` tags.

 For future reference here's an example of what worked for me:
```
<!-- wp:jetpack/markdown {\"source\":\"Hello\\\nWorld\"} -->\n<div class=\"wp-block-jetpack-markdown\"><p>Hello\\\nWorld</p>\n</div>\n<!-- /wp:jetpack/markdown -->
```
Like @etoledom said above, this isn't a good solution. I spent half a day yesterday digging into this but didn't make much progress. I understand that [JSON requires new line characters, `\n`, to be escaped](https://stackoverflow.com/questions/42068/how-do-i-handle-newlines-in-json#comment23184783_42073), resulting in `\\n`.
What I'm not sure about is whether the invalid content error we're seeing is only because of the newline, or if it also has something to do with the difference between what's in the JSON:
```
Hello\\nWorld
```
and what's in the `<p>` tags:
```
Hello\nWorld
```
It appears these have to be _exactly_ the same string for the invalid content message to not appear.

**Pull request merged**